Add Intel SGX support to Unikraft

xymeng16 opened the pull request:
### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [ ] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): `x86_64`
 - Platform(s): `kvm`
 - Application(s): N/A


### Additional configuration

 - `CONFIG_ARCH_X86_64_SGX=y`
 - `CONFIG_KVM_RING3=1`


### Description of changes

This PR is for adding Intel SGX support to Unikraft. It also introduces some tiny modifications on the GDT/TSS part to provide experimental ring-3 support, because SGX ENCLU instruction must be executed under ring-3.

### Introduction

This library `uksgx` provides the SGX driver, SDK and platform software (PSW, aka runtime), to provide a basic SGX functionality to Unikraft, which allows an enclave in the CPU to be executed securely at native speed. The enclave memory is not encrypted and can only be accessed by the enclave itself, guaranteed by the MMU inside the CPU package. Enclave runs at ring-3 only, hence no system calls are allowed. The enclave is linked statically and is self-contained with the pre-compiled SDK libraries.

A device `/dev/sgx` is registered to the devfs with the following IOCTL commands (for compatibility consideration):
- SGX_IOC_ENCLAVE_CREATE: validate user parameters and invoke ECREATE instruction
- SGX_IOC_ENCLAVE_ADD_PAGE: invoke EADD to add REG/TCS pages into the enclave and invoke EEXTEND to generate cryptographic hashes of the contents in the enclave
- SGX_IOC_ENCLAVE_INIT: verify enclave's content against signed SIGSTRUCT and initialize the enclave, then mark it ready to be used via EINIT instruction

After creation/loading/initialization, app/enclave can invoke EENTER/EEXIT/ERESUME to control the in/out of the enclave. Those switches are handled by `sgx_ocall()` and `sgx_ecall()`, which is wrapped in the SDK.

### Timeline
It is only a drafted PR and lots of work is required to be done. At the final time, I will squash the commits and do some cleanup before submitting the formal PR.
- 11 Jun 2022: At the time of the first draft, the probing of SGX is finished, that is to say, Unikraft can be aware of SGX if enabled in `make menuconfig`. And as inspired by [linux-sgx-driver](https://github.com/intel/linux-sgx-driver), the uksgx library registers a device /dev/sgx to the devfs, and some stub ioctl function is added to the library, in order to provide a unified and easy interaction channel between uksgx and the applications.
- 23 Jun 2022: As of today, the experimental ring-3 support dedicated to the SGX driver is finished when paging is enabled. Both paging and ring-3 are the fundamental features for SGX support.
- 7 Jul 2022: The EPC physical area is successfully mapped to the virtual address area given a hardcoded `SGX_VA_OFFSET` as its beginning address. This should be refined when the function for finding a usable VA area is implemented.
- 25 Jul 2022: `sgx_encl_create` is finished. Seems VMA is not available to Unikraft. So we need to interact with the page table using raw APIs for page management.
- 16 Aug 2022: #520 is online. This PR introduces a radix-tree implementation, which is required by SGX driver to manage the EPC pages.
- 8 Sep 2022: As of commit `ea1e61e`, the SGX driver is ported, except for implementations of a. VMA management; b. interruptible sleep and c. synchronized EPC page adding. Those, especially for the VMA management required a thorough review and understanding of the Unikraft's memory implementation, which requires lots of time. Hence I am going to work on the SGX SDK first.
- 14 Sep 2022: 
  1.  SDK is ported to Unikraft. Once `uksgx` is enabled, the SGX SDK will be compiled to some static libraries (`libsgx_tstdc.a`, `libsgx_tcxx.a` and `libsgx_tservice.a`) by default.
  2. Update this PR to add a basic introduction of Intel SGX implementation in Unikraft

xymeng16 committed:
arch: Add SGX related config to menuconfig.

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
arch/x86: Refactor the structure of SGX menuconfig.

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: Add uksgx and implement sgx_probe()

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
arch/x86: Refactor the structure of SGX menuconfig.

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: Remove `static` to fix link error

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: implements sgx_add_epc_bank and sgx_init

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
Merge branch 'unikraft:staging' into xiangyi/sgx

xymeng16 committed:
lib/uksgx: implementing sgx_reset_pubkey_hash

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: sgx_reset_pubkey_hash seems only work with SGX_LC==true, need to test

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: check if flexible launch control is supported

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: add some ioctl related functions

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: register /dev/sgx to devfs

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: configure /dev/sgx open/close/ioctl

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: working on ioctl-related code

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: creating dummy user-mode

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
plat/kvm/x86: implementing ring3 support for sgx

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: implementing switch_to_ring3 due to sgx requirements

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
Merge branch 'xiangyi/user-mode' into xiangyi/sgx

xymeng16 committed:
Merge branch 'unikraft:staging' into xiangyi/sgx

xymeng16 committed:
Merge branch 'unikraft:staging' into xiangyi/user-mode

xymeng16 committed:
Merge branch 'xiangyi/sgx' into xiangyi/user-mode

xymeng16 committed:
lib/uksgx: use iretq to switch to ring-3 (to be tested)

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
Merge branch 'xiangyi/sgx' into xiangyi/user-mode

xymeng16 committed:
plat/kvm/x86: fixed user mode GPT descriptors

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: switch_to_ring3() receive strange SIGQUIT

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
Merge branch 'unikraft:staging' into xiangyi/sgx

xymeng16 committed:
plat/kvm/x86: fix switch_to_ring3 iretq problem

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
Merge branch 'xiangyi/user-mode' into xiangyi/sgx

xymeng16 committed:
Merge branch 'unikraft:staging' into xiangyi/sgx

xymeng16 committed:
plat/kvm/x86: exception handler works normally under ring 3

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
plat/kvm/x86: modify ring-3 GDT descriptor index to enable
syscall/sysret fast CPL switch

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: refactor switch_to_ring3()
As we involved fast_cpl_switch() that leverages syscall/sysret
instructions to achieve fast CPL switch, the old switch_to_ring3()
is refactored into several functions, taking charge of initialization/
legacy/fast, seperately.

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
Merge branch 'unikraft:staging' into xiangyi/sgx

xymeng16 committed:
plat/kvm/x86: remove redundant comments

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: change to use macro in cpl_switch()

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: remove legacy_cpl_switch()
legacy_cpl_switch() uses iretq and int 80h to implement ring-0/ring0-3
switch, which are deprecated by hitsory. As Unikraft only runs on
x86_64 CPU, there is no reason to provide such fallback function
as modern x86_64 CPUs must support syscall/sysret instructions.

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
Merge branch 'unikraft:staging' into xiangyi/sgx

xymeng16 committed:
Merge branch 'unikraft:staging' into xiangyi/sgx

xymeng16 committed:
lib/uksgx: make the definition of cpl_switch_init() consistent

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
Merge branch 'unikraft:staging' into xiangyi/sgx

xymeng16 committed:
Merge branch 'unikraft:staging' into xiangyi/sgx

xymeng16 committed:
lib/uksgx: refactor wrmsr and add encls macro

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
Merge branch 'xiangyi/sgx' of github.com:xymeng16/unikraft into xiangyi/sgx

xymeng16 committed:
lib/uksgx: add encls C interface, working on paging

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
plat/x86: add ring3 support when paging is enabled

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: add physical-virtual addr mapping for SGX memory

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: add virtual address debug info

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: map EPC physical address to virtual address space

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: enforce SGX page size as 4K

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

cristian-vijelie commented on lib/uksgx/include/uk/sgx_cpu.h:
```diff
@@ -0,0 +1,283 @@
+#define ineI 0x49656e69
+#define ntel 0x6c65746e
+
+static inline void rdmsr(unsigned int msr, __u32 *lo, __u32 *hi)
```
You can use the functions already defined in `plat/common/include/x86/cpu.h`

Same for `rdmsrl`, `wrmsr`, `wrmsrl`

cristian-vijelie commented on lib/uksgx/include/uk/sgx_cpu.h:
```diff
@@ -0,0 +1,283 @@
+		     : "a"(y), "c"(z))
+
+/* __cpuid(unsinged int info[4], unsigned int leaf); */
+#define __cpuid(x, y)                                                    \
```
You can use the `cpuid` function from `plat/common/include/x86/cpu.h`

cristian-vijelie commented on lib/uksgx/include/uk/sgx_cpu.h:
```diff
@@ -0,0 +1,283 @@
+ */
+extern unsigned long __force_order;
+
+static inline unsigned long native_read_cr0(void)
```
Maybe it would be a good idea to define the functions related to the x86 architecture in `plat/common/include/x86/cpu.h`, so we don't have duplicates

xymeng16 replied:
Yep, I defined them in my own lib because at first I am not familiar with other parts of Unikraft hence do not want to modify it. I will perform a simplification and clarification once we reach a great milestone of this project, say make a "hello world" work from SGX.

xymeng16 committed:
lib/uksgx: working on sgx_ioc_enclave_create

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
Merge branch 'unikraft:staging' into xiangyi/sgx

xymeng16 committed:
lib/uksgx: `sgx_alloc_page` finished, working on `sgx_encl_create`

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
plat/kvm/x86: sync setup.c with original repo

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
plat/kvm/x86: sync setup.c with original repo

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: sgx_init_page finished, working on sgx_alloc_va_slot

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: remove useless Linux-specific marco

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: sgx_find_tgid_ctx, working on tgid issue

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: `sgx_encl_create` finished (VMA not impl yet)

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: working on sgx_encl_add_page
linux-sgx-driver handles adding page requests asynchronously,
but in unikraft, work queue is not implemented. Hence for this
version of implementation, the request will be handled once issued.

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
include/uk: working on radix tree.

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
move radix_tree.h to a seperate branch

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
Merge branch 'unikraft:staging' into xiangyi/sgx

xymeng16 committed:
include/uk/radix-tree: import linux radix-tree code from freebsd

Original file <commit 307f78f3ed90a4145eeb2c8cc79bc95b2666f57a>:
    - sys/compat/linuxkpi/common/include/linux/radix-tree.h
    - sys/compat/linuxkpi/common/src/linux_radix.c

This commit import linux radid-tree from freebsd source code.
Modifications are applied to the original code to make it adapted
to the Unikraft's API. This version does not introduce any lock to
the data structure, hence users should take care of the
synchronization by themselves.

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 commented:
#520 introduce the radix-tree implementation to Unikraft. This is required by SGX driver to manage the enclave pages. Once it passes the tests, I will merge it into the SGX driver branch.

xymeng16 committed:
include/uk/radix-tree: add some comments

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
Merge branch 'xiangyi/radix' into xiangyi/sgx

xymeng16 committed:
lib/uksgx: Enable radix_tree in SGX driver.

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: ioctl functions finished

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
include/uk/radix-tree: make functions `static inline`

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
Merge branch 'xiangyi/radix' into xiangyi/sgx

xymeng16 committed:
lib/uksignal: export `uk_sig_th_get_pending_any` for uksgx

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: add psw and sdk code

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx/sdk: add Makefile.uk for tlibthread

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx/sdk: figure out the structure of the sdk and psw

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: add simd headers and fix compile process

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

xymeng16 committed:
lib/uksgx: involve SGX SDK into Unikraft's compile system

Signed-off-by: Xiangyi Meng <xymeng16@gmail.com>

hlef commented:
Thank you so much for this PR @xymeng16!

A few high-level comments and points that I (we) came to realize since we started this project:
- If we take a step back, what is our threat model here? Using SGX within Unikraft, we essentially do not trust (some of the) "kernel" part of Unikraft. I don't know how much sense this makes; this part is already very thin! Clearly, it does bring a little something, but is it worth the performance, compatibility, and effort?
- More specifically, SGX is probably not the best target mechanism for Unikraft / unikernels in general. It forces you to switch to ring 3, a very inelegant and costly design quirk. Compared to other mechanisms that we can use for compartmentalization within Unikraft (MPK, etc.), it doesn't feel like a good match.
- Finally, SGX is a dying technology. It has been deprecated on laptop CPUs and it feels a bit like its existence on server processors is only a matter of time ([interesting discussion](https://github.com/intel/linux-sgx/issues/760)). Beyond SGX, the industry is moving away from the SGX model for a confidential VM model (AMD SEV, Intel TDX) which is much more suited for Unikraft and unikernels in general.

So, nothing against your work @xymeng16, it's great, we realized a lot of things! :+1: 

That being said, I do not think that we should ultimately merge this PR.

Comments?

dimakuv commented:
Hello, a Gramine maintainer here. I accidentally stumbled upon this PR.

> It forces you to switch to ring 3, a very inelegant and costly design quirk. Compared to other mechanisms that we can use for compartmentalization within Unikraft (MPK, etc.), it doesn't feel like a good match.

Intel SGX technology has nothing to do with compartmentalization. It's about confidential computing (so that privileged software and hardware adversaries cannot read/modify your application's memory). Comparing SGX vs MPK is wrong.

> Finally, SGX is a dying technology. [...] it feels a bit like its existence on server processors is only a matter of time.

That's incorrect.
